File: optionals/med-4.0.0/tests/c/2.3.6/test13.c
/*  This file is part of MED.
 *
 *  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
 *  MED is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  MED is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with MED.  If not, see <http://www.gnu.org/licenses/>.
 */

/******************************************************************************
 * - Nom du fichier : test13.c
 *
 * - Description : lecture des equivalences d'un maillage MED.
 *
 *****************************************************************************/

#include <med.h>
#define MESGERR 1
#include <med_utils.h>

#ifdef DEF_LECT_ECR
#define MODE_ACCES MED_LECTURE_ECRITURE
#elif DEF_LECT_AJOUT
#define MODE_ACCES MED_LECTURE_AJOUT
#else
#define MODE_ACCES MED_CREATION
#endif

/* On prend en compte tous les types de mailles concernees
 *    par les equivalences */
#define MED_NBR_MAILLE_EQU 8
 
int main (int argc, char **argv)


{
  med_err ret = 0;
  med_idt fid;
  char maa[MED_TAILLE_NOM+1];
  med_int mdim;
  med_int nequ,ncor;
  med_int *cor;
  char equ[MED_TAILLE_NOM+1];
  char des[MED_TAILLE_DESC+1];
  med_geometrie_element typmai[MED_NBR_GEOMETRIE_MAILLE+1] = {MED_POINT1,MED_SEG2, 
							      MED_SEG3,MED_TRIA3,
							      MED_TRIA6,MED_QUAD4,
							      MED_QUAD8,MED_POLYGONE};
  med_geometrie_element typfac[MED_NBR_GEOMETRIE_FACE+1] = {MED_TRIA3,MED_TRIA6,
							    MED_QUAD4,MED_QUAD8,
							    MED_POLYGONE};
  med_geometrie_element typare[MED_NBR_GEOMETRIE_ARETE] = {MED_SEG2,MED_SEG3};  
  int i,j,k;
  med_maillage type;

  if (argc != 2) {
    MESSAGE("Il faut passer un fichier MED en paramètre");
    return -1;
  }

  /* Ouverture du fichier passe en argument en lecture seule */
  if ((fid = MEDouvrir(argv[1],MED_LECTURE)) < 0) {
    MESSAGE("Erreur a l'ouverture du fichier : "); SSCRUTE(argv[1]);
    return -1;
  }
  
  /* Lecture des infos sur le premier maillage */
  if (MEDmaaInfo(fid,1,maa,&mdim,&type,des) < 0) {
    MESSAGE("Erreur a lecture des infos sur le 1er maillage"); 
    return -1;
  }
  printf("Maillage de nom %s et de dimension "IFORMAT" \n",maa,mdim);

  /* Lecture du nombre d'equivalence */
  if ((nequ = MEDnEquiv(fid,maa)) < 0) {
    MESSAGE("Erreur a la lecture du nombre d'equivalence"); 
    return -1;
  }
  printf("Nombre d'equivalences : "IFORMAT" \n",nequ);

  /* Lecture de toutes les equivalences du maillage */
  if (nequ > 0)
    for (i = 0;i<nequ;i++) {
      printf("Equivalence numero : %d \n",i+1);

      /* Lecture des infos sur l'equivalence */
      if (MEDequivInfo(fid,maa,i+1,equ,des) < 0) {
	MESSAGE("Erreur a la lecture de l'equivalence d'indice");
	ISCRUTE_int(i+1);
	return -1;
      }
      printf("Nom de l'equivalence: %s \n",equ);
      printf("Description de l'equivalence : %s \n",des);

      /* Lecture des correspondances sur les differents types d'entites */

      /* Les noeuds */
      if ((ncor = MEDnCorres(fid,maa,equ,MED_NOEUD,0)) < 0) {
	MESSAGE("Erreur a la lecture du nombre de correspondance sur les noeuds");
	return -1;
      }
      printf("Il y a "IFORMAT" correspondances sur les noeuds \n",ncor);
      if (ncor > 0) {
	cor = (med_int*) malloc(sizeof(med_int)*ncor*2);
	if (MEDequivLire(fid,maa,equ,cor,ncor,MED_NOEUD,0) < 0) {
	  MESSAGE("Erreur a la lecture des correspondances sur les noeuds");
	  ret = -1;
	}
	if (ret == 0)
	  for (j=0;j<ncor;j++)
	    printf("Correspondance %d : "IFORMAT" et "IFORMAT" \n",j+1,*(cor+2*j),
		   *(cor+2*j+1));
	free(cor);
      }
	    
      /* Les mailles : on ne prend pas en compte les mailles 3D */
      if (ret == 0)
	for (j=0;j<MED_NBR_MAILLE_EQU;j++) {

	  if ((ncor = MEDnCorres(fid,maa,equ,MED_MAILLE,typmai[j])) < 0) {
	    MESSAGE("Erreur a la lecture du nombre de correspondance sur les mailles : ");
	    ISCRUTE_int(typmai[j]);
	    return -1;
	  }
	  printf("Il y a "IFORMAT" correspondances sur les mailles %d \n",ncor,
		 typmai[j]);
	  if (ncor > 0) {
	    cor = (med_int*) malloc(sizeof(med_int)*ncor*2);
	    if (MEDequivLire(fid,maa,equ,cor,ncor,MED_MAILLE,
			     typmai[j]) < 0) {
	      MESSAGE("Erreur a la lecture des correspondances sur les mailles : ");
	      ISCRUTE_int(typmai[j]);
	      ret = -1;
	    }
	    if (ret == 0)
	      for (k=0;k<ncor;k++)
		printf("Correspondance %d : "IFORMAT" et "IFORMAT" \n",k+1,*(cor+2*k),
		       *(cor+2*k+1));
	    free(cor);
	  }
	}

      /* Les faces */
      if (ret == 0)
	for (j=0;j<MED_NBR_GEOMETRIE_FACE+1;j++) {
	  if ((ncor = MEDnCorres(fid,maa,equ,MED_FACE,typfac[j])) < 0) {
	    MESSAGE("Erreur a la lecture du nombre de correspondance sur les faces : ");
	    ISCRUTE_int(typfac[j]);
	    return -1;
	  }
	  printf("Il y a %d correspondances sur les faces "IFORMAT" \n",ncor,
		 typfac[j]);
	  if (ncor > 0) {
	    cor = (med_int*) malloc(sizeof(med_int)*ncor*2);
	    if (MEDequivLire(fid,maa,equ,cor,ncor,MED_FACE,
			     typfac[j]) < 0) {
	      MESSAGE("Erreur a la lecture des correspondances sur les faces : ");
	      ISCRUTE_int(typfac[j]);
	      ret = -1;
	    }
	    if (ret == 0)  
	      for (k=0;k<ncor;k++)
		printf("Correspondance %d : "IFORMAT" et "IFORMAT" \n",k+1,*(cor+2*k),
		       *(cor+2*k+1));
	    free(cor);
	  }
	}

      /*  Les aretes */
      if (ret == 0)
	for (j=0;j<MED_NBR_GEOMETRIE_ARETE;j++) {
	  if ((ncor = MEDnCorres(fid,maa,equ,MED_ARETE,typare[j])) < 0) {
	    MESSAGE("Erreur a la lecture du nombre de correspondance sur les aretes : ");
	    ISCRUTE_int(typare[j]);
	    return -1;
	  }
	  printf("Il y a "IFORMAT" correspondances sur les aretes %d \n",ncor,
		 typare[j]);
	  if (ncor > 0) {
	    cor = (med_int*) malloc(sizeof(med_int)*ncor*2);
	   if (MEDequivLire(fid,maa,equ,cor,ncor,MED_ARETE,
			       typare[j]) < 0) {
	     MESSAGE("Erreur a la lecture des correspondances sur les faces : ");
	     ISCRUTE_int(typare[j]);
	     ret = -1;
	   }
	   if (ret == 0)
	     for (k=0;k<ncor;k++)
	       printf("Correspondance %d : "IFORMAT" et "IFORMAT" \n",k+1,*(cor+2*k),
		      *(cor+2*k+1));
	   free(cor);
	  }
	}

    }			    

  /* Fermeture du fichier */
  if (MEDfermer(fid) < 0) {
    MESSAGE("Erreur a la fermeture du fichier ");
    return -1;
  }

  return ret;
}

File: optionals/med-4.0.0/tests/f/2.3.6/test13.f90
!*  This file is part of MED.
!*
!*  COPYRIGHT (C) 1999 - 2019  EDF R&D, CEA/DEN
!*  MED is free software: you can redistribute it and/or modify
!*  it under the terms of the GNU Lesser General Public License as published by
!*  the Free Software Foundation, either version 3 of the License, or
!*  (at your option) any later version.
!*
!*  MED is distributed in the hope that it will be useful,
!*  but WITHOUT ANY WARRANTY; without even the implied warranty of
!*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
!*  GNU Lesser General Public License for more details.
!*
!*  You should have received a copy of the GNU Lesser General Public License
!*  along with MED.  If not, see <http://www.gnu.org/licenses/>.
!*

! ******************************************************************************
! * - Nom du fichier : test13.f90
! *
! * - Description : lecture des equivalences dans un maillage MED.
! *
! ******************************************************************************

program test13
  
  implicit none
  include 'med.hf'
!
!
  integer*8    fid
  integer      ret,cret
  character*32 maa
  integer      mdim,nequ,ncor
  integer, allocatable, dimension(:) :: cor
  character*32  equ
  character*200 des
  integer       i,j,k
  character*255 argc
  integer, parameter :: MED_NBR_MAILLE_EQU = 8
  integer,parameter  :: typmai(MED_NBR_MAILLE_EQU) =  (/ MED_POINT1,MED_SEG2,   &
                                                      &  MED_SEG3,MED_TRIA3,    &
                                                      &  MED_TRIA6,MED_QUAD4,   &
                                                      &  MED_QUAD8,MED_POLYGONE/)

   integer,parameter :: typfac(MED_NBR_GEOMETRIE_FACE+1) = (/MED_TRIA3,MED_TRIA6,       &
					&	 MED_QUAD4,MED_QUAD8, MED_POLYGONE/)
   integer,parameter ::typare(MED_NBR_GEOMETRIE_ARETE) = (/MED_SEG2,MED_SEG3/)
   character*200 desc
   integer type
 
   print *,"Indiquez le fichier med a decrire : "
   !!read(*,*) argc
   argc = "test12.med"

   !  ** Ouverture du fichier en lecture seule **
   call efouvr(fid,argc,MED_LECTURE, cret)
   print *,cret
   
     
   !  ** Lecture des infos sur le premier maillage **
   if (cret.eq.0) then
      call efmaai(fid,1,maa,mdim,type,desc,cret)
      print *,"Maillage de nom : ",maa," et de dimension : ", mdim
   endif
   print *,cret


   !  ** Lecture du nombre d'equivalence  **
   if (cret.eq.0) then
      call efnequ(fid,maa,nequ,cret)
      if (cret.eq.0) then
         print *,"Nombre d'equivalences : ",nequ
      endif
   endif
 
   !** Lecture de toutes les equivalences **
   if (cret.eq.0) then
      do i=1,nequ
         print *,"Equivalence numero : ",i
         !** Lecture des infos sur l'equivalence **
         if (cret.eq.0) then
            call efequi(fid,maa,i,equ,des,cret)
         endif
         print *,cret
         if (cret.eq.0) then
            print *,"Nom de l'equivalence : ",equ          
            print *,"Description de l'equivalence : ",des 
         endif

         !** Lecture des correspondances sur les differents types d'entites **
         if (cret.eq.0) then
	    !** Les noeuds **
            call efncor(fid,maa,equ,MED_NOEUD,0,ncor,cret)
            print *,"Il y a ",ncor," correspondances sur les noeuds "
            if (ncor > 0) then
               allocate(cor(ncor*2),STAT=ret)
               call efequl(fid,maa,equ,cor,ncor,MED_NOEUD,0,cret)
               do j=0,(ncor-1)
                  print *,"Correspondance ",j+1," : ",cor(2*j+1)," et ",cor(2*j+2)
               end do
               deallocate(cor)
            end if
	    
	    !** Les mailles : on ne prend pas en compte les mailles 3D **

            do j=1,MED_NBR_MAILLE_EQU
               call efncor(fid,maa,equ,MED_MAILLE,typmai(j),ncor,cret)
               print *,"Il y a ",ncor," correspondances sur les mailles ",typmai(j)
               if (ncor > 0 ) then
                  allocate(cor(2*ncor),STAT=ret)
                  call efequl(fid,maa,equ,cor,ncor,MED_MAILLE,typmai(j),cret)
                  do k=0,(ncor-1)
                     print *,"Correspondance ",k+1," : ",cor(2*k+1)," et ",cor(2*k+2)
                  end do
                  deallocate(cor)
               endif
            end do

	    ! ** Les faces **
            do j=1,MED_NBR_GEOMETRIE_FACE+1
               call efncor(fid,maa,equ,MED_FACE,typfac(j),ncor,cret)
               print *,"Il y a ",ncor," correspondances sur les faces ",typfac(j)
               if (ncor > 0 ) then
                  allocate(cor(2*ncor),STAT=ret)
                  call efequl(fid,maa,equ,cor,ncor,MED_FACE,typfac(j),cret)
                  do k=0,(ncor-1)
                     print *,"Correspondance ",k+1," : ",cor(2*k+1)," et ",cor(2*k+2)
                  end do
                  deallocate(cor)
               endif
            end do

	    ! **  Les aretes **
            do j=1,MED_NBR_GEOMETRIE_ARETE
               call efncor(fid,maa,equ,MED_ARETE,typare(j),ncor,cret)
               print *,"Il y a ",ncor," correspondances sur les aretes ",typare(j)
               if (ncor > 0 ) then
                  allocate(cor(2*ncor),STAT=ret)
                  call efequl(fid,maa,equ,cor,ncor,MED_ARETE,typare(j),cret)
                  do k=0,(ncor-1)
                     print *,"Correspondance ",k+1," : ",cor(2*k+1)," et ",cor(2*k+2)
                  end do
                  deallocate(cor)
               endif
            end do

         end if
      end do
   end if

!  ** Fermeture du fichier   **
   call efferm (fid,cret)
   print *,cret

!  ** Code retour
   call efexit(cret)
   
 end program test13
	




